magma/primitives/memory.py

```diff
@@ -106,3 +110,25 @@ def __init__(self, height, T: Kind,
         for _ in range(read_latency - 1):
             rdata = Register(T)()(rdata)
         self.io.RDATA @= rdata
+
+        def __setitem__(self, addr, data):
+            if self.WADDR.driven() or self.WDATA.driven():
+                _logger.warning(
+                    "Calling __setitem__ on a Memory instance with WADDR or"
+                    " WDATA already driven, will overwrite previous values"
+                )
+            self.WADDR @= addr
+            self.WDATA @= data
+
+        self.__setitem__ = __setitem__
+
+        def __getitem__(self, addr):
+            if self.RADDR.driven():
+                _logger.warning(
+                    "Calling __getitem__ on a Memory instance with RADDR"
+                    " already driven, will overwrite previous values"
+                )
+            self.RADDR @= addr
+            return self.RDATA
+
+        self.__getitem__ = __getitem__
```

Comment:
Wasn't too happy with how this had to be done, I wonder if we can come up with something better for defining methods on generated Circuits.  Perhaps we could define the methods inside the Generator2 subclass body, and mark them as `@circuitmethod` or something like that?

Reply:
If I understand correctly the issue is having member functions on the the instances? And right now a `def foo(self)` would be a member function on the module (/circuit). So the only way around it is to do `self.foo = fn`. Just trying to make sure I understand.

I think a decorator would work well. Let me think about it. It's kind of a weird problem and probably where the "elegance" of the solution breaks down.

Reply:
Understanding is correct (defining a method here defines it on the class instead of the instance)

Reply:
Ok. I have an idea of how to do it similar to your idea. If you want to merge this, I can make a PR shortly that adds this feature and updates this memory generator to use it.